addons/auth_totp_mail_enforce/models/res_users.py

```diff
@@ -0,0 +1,139 @@
+        code = str(int(hm, 16))[-6:]
+        expiration = now + timedelta(seconds=3600)
+        return code, expiration
+
+    def _validate_totp_mail_code(self, code):
+        if not code:
+            return False
+
+        token = f'{self.id}/{self.login}'
+        secret = self.env['ir.config_parameter'].sudo().get_param('database.secret')
+        assert secret, "Mail TOTP protection requires a configured database secret"
+        t = time.time()
+        low = int((t - 3600) / 3600)
+        high = int((t + 3600) / 3600) + 1
+        for counter in range(low, high):
+            msg = '%s%s' % (token, counter)
+            hm_expected = hmac.new(secret.encode('ascii'), msg.encode('utf-8'), hashlib.sha1).hexdigest()
+            if code == str(int(hm_expected, 16))[-6:]:
+                return True
+        return False
+
+    def _send_totp_mail_code(self, user_demand=False):
+        self.ensure_one()
+        self._totp_rate_limit('send_email')
+
+        template = self.env.ref('auth_totp_mail_enforce.mail_template_totp_mail_code').sudo()
+        odoobot = self.env.ref('base.partner_root').sudo()
+        email_values = {
+            'email_from': odoobot.email_formatted,
+            'author_id': odoobot.id,
+        }
+        context = {}
+        if request:
+            geoip = request.session.geoip
+            # e.g.
+            # from 'Mozilla/5.0 (Linux; Android 11; Pixel 5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.91'
+            # get 'Mozilla/5.0' as browser and 'Linux; Android 11; Pixel 5' as device
+            [(browser, device)] = re.findall('^([^ ]*)(?: \\(([^\\(]*)\\))?', request.httprequest.environ['HTTP_USER_AGENT'])
+            context.update({
+                'location': f"{geoip['city']}, {geoip['country_name']}" if geoip else None,
+                'device': device,
+                'browser': browser,
+                'ip': request.httprequest.environ['REMOTE_ADDR'],
+            })
+        return template.with_context(**context).send_mail(
+            self.id, force_send=True, email_values=email_values, notif_layout='mail.mail_notification_light'
+        )
+
+    def _totp_rate_limit(self, limit_type):
+        self.ensure_one()
+        assert request, "A request is required to be able to rate limit TOTP related actions"
+        ip = request.httprequest.environ['REMOTE_ADDR']
+        description, limit, interval = TOTP_RATE_LIMITS.get(limit_type)
+        RateLimitLog = self.env['auth.totp.rate.limit.log'].sudo()
+        domain = [
+            ('user_id', '=', self.id),
+            ('create_date', '>=', datetime.now() - timedelta(seconds=interval)),
+            ('limit_type', '=', limit_type),
+            ('ip', '=', ip),
+        ]
+        count = RateLimitLog.search_count(domain)
+        if count >= limit:
+            raise AccessDenied(_('You reached the limit of %s for your account', description))
```

Comment:
It's better to have 2 long messages "You have reached the limit of authentication email..." instead of mixing translatable parts. Some languages do not put the words in the same order and it would be very diffucult for them to translate.